go/vt/vtctl/workflow/server.go

```diff
@@ -2128,25 +2129,52 @@ func (s *Server) WorkflowDelete(ctx context.Context, req *vtctldatapb.WorkflowDe
 		s.optimizeCopyStateTable(tablet.Tablet)
 		return res.Result, err
 	}
-	delCtx, delCancel := context.WithTimeout(ctx, topo.RemoteOperationTimeout*2)
-	defer delCancel()
-	res, err := vx.CallbackContext(delCtx, callback)
-	if err != nil {
-		return nil, err
-	}
 
-	if len(res) == 0 {
-		return nil, vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "the %s workflow does not exist in the %s keyspace", req.Workflow, req.Keyspace)
+	var res map[*topo.TabletInfo]*querypb.QueryResult
+	delFunc := func() error {
+		delCtx, delCancel := context.WithTimeout(ctx, topo.RemoteOperationTimeout*2)
+		defer delCancel()
+		res, err = vx.CallbackContext(delCtx, callback)
+		if err != nil {
+			return err
+		}
+		return nil
 	}
 
-	// Cleanup related data and artifacts. There are none for a LookupVindex workflow.
-	if ts.workflowType != binlogdatapb.VReplicationWorkflowType_CreateLookupIndex {
-		if _, err := s.DropTargets(delCtx, ts, req.GetKeepData(), req.GetKeepRoutingRules(), false); err != nil {
-			if topo.IsErrType(err, topo.NoNode) {
-				return nil, vterrors.Wrapf(err, "%s keyspace does not exist", req.GetKeyspace())
+	// Multi-tenant migrations delete only that tenant's records from the target tables
+	// in batches and we may not be able to complete that work before the timeout. So
+	// we only delete the workflow after the cleanup work completes successfully so that
+	// the workflow can be canceled multiple times if needed in order to fully cleanup
+	// all of the tenant's data that we had copied.
+	if ts.IsMultiTenantMigration() {
+		if ts.workflowType != binlogdatapb.VReplicationWorkflowType_MoveTables { // Should never happen
+			return nil, vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "unsupported workflow type %q for multi-tenant migration",
+				ts.workflowType)
+		}
+		// We need to delete the rows that the target tables would have for the tenant.
+		// We don't cleanup other related artifacts since they are not tied to the tenant.
+		if !req.GetKeepData() {
+			if err := s.DeleteTenantData(ctx, ts, req.DeleteBatchSize); err != nil {
+				return nil, vterrors.Wrapf(err, "failed to fully delete all migrated data for tenant %s, please retry the operation",
+					ts.options.TenantId)
 			}
+		}
+		if err := delFunc(); err != nil {
 			return nil, err
 		}
+	} else {
+		if err := delFunc(); err != nil {
```

Comment:
I'd rather not do that here as deleting the workflow first was intentional as part of: https://github.com/vitessio/vitess/pull/15977 We can revisit that, but I'd rather not change it here as it's unrelated to this work.

Reply:
To be clear, I think it would be good to delete the workflow last in the non-multi-tenant case. But to do that, I think that we should also stop the workflow streams in that case before deleting the related artifacts, delete the artifacts, then delete the workflow. I would just rather not do that here as it's not related to the PR.

Reply:
Did this here: https://github.com/vitessio/vitess/pull/16906/commits/e7374c8575f5f0552034050c9980ad62b21ab568

Reply:
So... I changed my mind and decided the bite the bullet as I had the problem space clear in my head ATM 😄 https://github.com/vitessio/vitess/pull/16906/commits/6a4c79e69a2125b9467fb8d7b176c3d30152a92f